src/docs/services/relay/projectconfig-versioning.mdx

```diff
@@ -0,0 +1,138 @@
+
+```rust
+struct Foo {
+    foo: u64,
+}
+
+// compatible with: {"foo": 123}, {"foo": 123, "bar": 456}, {"foo": 123, ...}
+// incompatible with: {"foo": "123"}
+```
+
+There are two caveats:
+
+- As demonstrated above, changing types is *not* safe. See below for instructions on changing types.
+- Ensure the field name has never been used before. It could have been implemented with different types, or trigger unwanted behavior in prior versions of Relay.
+- Note that adding enum variants requires a different approach.
+
+## Renaming Fields
+
+Do not rename a field for features that are generally available for external Relays. Renaming a field makes the field invisible to older Relays. This makes it infeasible to rename fields for aesthetic reasons (it *may* be fine if your feature is pre-EA), but for the same reason [it is a useful mechanism for versioning](#breaking-changes-and-actual-versioning)
+
+## Changing Field Types
+
+This should be generally OK as long as it serializes/deserializes the same, for example you can swap out `String` with `smartstring::String`.
+
+Decreasing integer sizes (`u64` to `u32`) is a breaking change unless you can somehow prove all possible values fit into the new integer type, else it cannot be done without [renaming the field](#breaking-changes-and-actual-versioning). Serde throws an error if the value-to-deserialize does not fit into the integer type.
+
+Increasing integer sizes (`u32` to `u64`) should be fine in most cases. Just keep in mind that when you increase the integer size and deploy internal relays, you push potential overflow errors further to the edge (or customer) where you no longer see them.
+
+In case your field is emitted for external Relays, **changing a field type is rarely possible** without [renaming the field](#breaking-changes-and-actual-versioning) entirely. Even if you think it's feasible, renaming the field in addition is often a worthwhile additional security dance.
+
+In case it's not emitted to external relays (not part of `LimitedProjectConfig`), you may be able to at least extend the set of types Relay accepts. For that it's recommended to either:
+
+- [Use an untagged enum](https://serde.rs/enum-representations.html#untagged) (this is like `Union[str, int]` in Python)
+- Write a custom `Deserialize` impl for your type to be able to deserialize from older values
+
+## Deployment & Monitoring
+
+Talk to the ingest team for deploying and monitoring.
+
+## Breaking changes, and actual versioning
+
+> `myfield_v2_final_final.docx`
+
+We've mostly gone through things that we don't want you to do, as they will cause incidents.
+
+In case you do need to make larger changes to the schema, or you want to do a large amount of smaller changes and feel like guaranteeing backwards compat for all of them is too hard, consider renaming the top-level field `x` to `x_v2`.
```

Comment:
There is another approach to version the project config format. See https://github.com/getsentry/relay/blob/3933cd6ccd59c3270f56853ffbc8678b219e44e6/relay-server/src/actors/project_upstream.rs#L25-L34

Reply:
I now mention it as sidenote for *us* to investigate but I haven't thought enough about this to know when to use it. So I'd not write anything to steer people towards it